man/analyse.1
.TH "ANALYSE" "22" "Febuary 2010" "GENTOOLKIT" ""
.SH "NAME"
analyse \- Gentoo Installed Package Analysis Tool

.SH "SYNOPSIS"
.BI "analyse " "[global\-options] " "module " "[local\-options]" "TARGET"

.SH "DESCRIPTION"
.B Analyse
Is a collection of modules for analysing the state of installed Gentoo packages for
USE flags or keywords used for installation, and their current masking status.
.br 
It can also optionally (re)generate new /etc/portage/package.* files.
.br 

.br 
.B CAUTION:
This is beta software and is not yet feature complete. Some features, options or its name
may change in the future.   Any files that it generates are saved to your home directory
and will not harm your system without effort on your part.
.br 

.SH "GLOBAL OPTIONS"
.HP
.B \-h, \-\-help
.br 
Output a help message.
.HP
.B \-q, \-\-quiet
.br 
Be less verbose where possible. In some modules, this option can increase the output speed.
.HP
.B \-C, \-\-no\-color
.br 
Do not colorize output.
.HP
.B \-N, \-\-no\-pipe
.br 
Turn off automatic pipe detection. Use this option if you do not want
.B analyse
To detect if the output is being directed to the screen or to another program
and adjust color and verbosity accordingly.
.HP
.B \-V, \-\-version
.br 
Display \fBGentoolkit\fP's version. Please include this in all bug reports. (see
.B BUGS
below)

.SH "MODULES"
.B Analyse
Uses a system of modules. Each module has both a long and short name. 
The list below uses the notation "\fBmodule (m)\fP", where \fIm\fP is the short name
and \fImodule\fP is the long name.
.P
You can view the
.B help
message for a specific module by using
.BR "\-h" ", " "\-\-help "
as either a global option (after
.B analyse
and before the module name) or as a local option (after the module name).

.SS
.BI "analyse (a) [OPTIONS] TARGET"
Report on all installed packages for \fITARGET\fP.
.P

.I R "TARGET" ":"
.HP
.B use
.br 
Will analyse the installed with USE flags for output results.
.HP
.B pkguse
.br 
Will analyse the USE flags information from the installed pkg's 'PKGUSE' file which contains 
only flags settings from /etc/portage/package.use at the time of installation.
.HP
.B keywords
.br 
Will analyse the recorded keywords for output results.
.HP
.B unmask
.br 
Will analyse the installed packages and portage trees for pkgs that require unmasking and report them.
.br 
.P
.I R "LOCAL OPTIONS" ":"
.HP
.B \-u, \-\-unset
.br 
Will also include any USE flags used that were not enabled for some packages.
.HP
.B \-v, \-\-verebose
.br 
Gives more detail about the results found and the current task being performed.

.P
.I R "EXAMPLES" ":"
.EX
.HP
analyse a \-\-verbose \-\-unset use
.EE
.br 
Report on all use flags used to install the packages.  (\-\-unset) Include in the report all flags
that have been used but also were not set enabled for some packages.  
(\-\-verbose) Also list the packages that used the USE flag setting.
The report will break down the useage and report the USE flag up to 3 times indicating its
setting {"+","\-"," "= unset} prepended to the flag name. 
It will also color the output, red = Disabled, blue = Enabled, plain text = unset
.br 

.SS
.BI "rebuild (r) [OPTIONS] TARGET"
Create a list all packages for \fITARGET\fP settings that are needed for
other than the default settings.

.I R "TARGET" ":"
.HP
.B use
.br 
Will analyse the USE flags for output results.
.HP
.B keywords
.br 
Will analyse the keywords for output results.
.HP
.B unmask
.br 
Will analyse the installed packages and portage trees for pkgs that require
unmasking and produce output/a new /etc/portage/package.unmask file.
.P
.I R "LOCAL OPTIONS" ":"
.HP
.B \-a, \-\-all
.br 
Create files/output for all TARGET(s) found to need it. (not Implemented yet)
.HP
.B \-e, \-\-excact
.br 
Will prepend the pkg with = as well as use the version information for the entries.
.br 
eg.:  =CAT/PKG\-VER flag1 flag2
.HP
.B \-p, \-\-pretend
.br 
Sends the output to the screen instead of a file.
.HP
.B \-v, \-\-verebose
.br 
Gives more detail about the results found and the current task being performed.
.P
.I R "EXAMPLES" ":"
.EX
.HP
analyse rebuild \-p use
.EE
.br 
Analyse the installed packages database and current system USE flag settings
 and output the results in the form of:
.br 
   
.br 
.EX
CAT/PKG \-flag1 \-flag2 flag3 flag4...

.SS
.BI "clean (c) [OPTIONS] TARGET"
Clean all packages for \fITARGET\fP settings that are found with obsolete settings
for the current settings and pkg ebuild. (not Implemented yet)

.I R "TARGET" ":"
.HP
.B use
.br 
Will analyse the USE flags and /etc/portage/package.use file(s) for entries that
are redundant or no longer used by the pkg.
.HP
.B keywords
.br 
Will analyse the keywords and /etc/portage/package.keywords file(s) for entries
that are no longer needed.
.HP
.B unmask
.br 
Will analyse the installed packages, /etc/portage/package.unmask file(s) and
portage trees for pkgs that no longer require unmasking.
.P
.I R "LOCAL OPTIONS" ":"
.HP
.B \-a, \-\-all
.br 
Clean files/output for all TARGET(s) found to need it. (not Implemented yet)
.HP
.B \-p, \-\-pretend
.br 
Sends the output to the screen instead of a file.
.HP
.B \-v, \-\-verebose
.br 
Gives more detail about the results found and the current task being performed.


.SH "BUGS"
Submit bug reports to http://bugs.gentoo.org.

.SH "AUTHORS"
.br 
Brian Dolbec <brian.dolbec@gmail.com>, 2010


NEWS
News (new features/major bug fixes)

gentoolkit-0.3.0
----------------

epkginfo:
  * is now a link to equery meta and has all the features equery meta has

equery:
  * --help menus cleaned up, using notations (b)elongs instead of belongs(b).
  * man page has been rewritten.
  * --quiet effects more modules.
  * 2 new modules:
    * changes - Gentoo ChangeLog viewer, try:
      `equery changes portage` to see entry for portage version that emerge
        wants to install;
      `equery changes portage --from=2.2_rc20 --to=2.2_rc30` to see all entries
        between the specified versions.
    * meta - Displays information available in metadata.xml and keyword info.
      Try `equery meta boost` to list herd, maintainers, keywords, and more.
  * Modules which are meant to run on multiple packages (check, list, size) now
    allow category and package name globbing, (so no more need for --exact-name
    or --category).

      # Exact name matching by default:
      $ equery l zilla
       * Searching for zilla ...

      # Use globs to fuzzy match
      $ equery l *zilla*
       * Searching for *zilla* ...
      [IP-] [  ] www-client/mozilla-firefox-3.5.4:0

      # Use globs to 'category filter'
      $ equery l www-client/*
       * Searching for * in www-client ...
      [I--] [XX] www-client/chromium-4.0.223.5:0
      [IP-] [  ] www-client/epiphany-2.26.3-r2:0
      [IP-] [  ] www-client/links-2.2:2
      [IP-] [  ] www-client/mozilla-firefox-3.5.4:0


pym/gentoolkit/analyse/analyse.py
#!/usr/bin/python
#
# Copyright 2010 Brian Dolbec <brian.dolbec@gmail.com>
# Copyright(c) 2010, Gentoo Foundation
# Distributed under the terms of the GNU General Public License v2
#

"""Provides a breakdown list of USE flags or keywords used and by
what packages according to the Installed package database"""

from __future__ import print_function

import sys

import gentoolkit
from gentoolkit.dbapi import PORTDB, VARDB
from gentoolkit.analyse.base import ModuleBase
from gentoolkit import pprinter as pp
from gentoolkit.flag import get_installed_use, get_flags
from gentoolkit.analyse.lib import FlagAnalyzer, KeywordAnalyser
from gentoolkit.analyse.output import nl, AnalysisPrinter
from gentoolkit.package import Package
from gentoolkit.helpers import get_installed_cpvs

import portage


def gather_flags_info(
		cpvs=None,
		system_flags=None,
		include_unset=False,
		target="USE",
		use_portage=False,
		#  override-able for testing
		_get_flags=get_flags,
		_get_used=get_installed_use
		):
	"""Analyse the installed pkgs USE flags for frequency of use

	@type cpvs: list
	@param cpvs: optional list of [cat/pkg-ver,...] to analyse or
			defaults to entire installed pkg db
	@type: system_flags: list
	@param system_flags: the current default USE flags as defined
			by portage.settings["USE"].split()
	@type include_unset: bool
	@param include_unset: controls the inclusion of unset USE flags in the report.
	@type target: string
	@param target: the environment variable being analysed
			one of ["USE", "PKGUSE"]
	@type _get_flags: function
	@param _get_flags: ovride-able for testing,
			defaults to gentoolkit.analyse.lib.get_flags
	@param _get_used: ovride-able for testing,
			defaults to gentoolkit.analyse.lib.get_installed_use
	@rtype dict. {flag:{"+":[cat/pkg-ver,...], "-":[cat/pkg-ver,...], "unset":[]}
	"""
	if cpvs is None:
		cpvs = VARDB.cpv_all()
	# pass them in to override for tests
	flags = FlagAnalyzer(system_flags,
		filter_defaults=False,
		target=target,
		_get_flags=_get_flags,
		_get_used=get_installed_use
	)
	flag_users = {}
	for cpv in cpvs:
		if cpv.startswith("virtual"):
			continue
		if use_portage:
			plus, minus, unset = flags.analyse_cpv(cpv)
		else:
			pkg = Package(cpv)
			plus, minus, unset = flags.analyse_pkg(pkg)
		for flag in plus:
			if flag in flag_users:
				flag_users[flag]["+"].append(cpv)
			else:
				flag_users[flag] = {"+": [cpv], "-": []}
		for flag in minus:
			if flag in flag_users:
				flag_users[flag]["-"].append(cpv)
			else:
				flag_users[flag] = {"+":[], "-": [cpv]}
		if include_unset:
			for flag in unset:
				if flag in flag_users:
					if "unset" in flag_users[flag]:
						flag_users[flag]["unset"].append(cpv)
					else:
						flag_users[flag]["unset"] = [cpv]
				else:
					flag_users[flag] = {"+": [], "-": [], "unset": [cpv]}
	return flag_users


def gather_keywords_info(
		cpvs=None,
		system_keywords=None,
		use_portage=False,
		#  override-able for testing
		keywords=portage.settings["ACCEPT_KEYWORDS"],
		analyser = None
		):
	"""Analyse the installed pkgs 'keywords' for frequency of use

	@param cpvs: optional list of [cat/pkg-ver,...] to analyse or
			defaults to entire installed pkg db
	@param system_keywords: list of the system keywords
	@param keywords: user defined list of keywords to check and report on
			or reports on all relevant keywords found to have been used.
	@param _get_kwds: overridable function for testing
	@param _get_used: overridable function for testing
	@rtype dict. {keyword:{"stable":[cat/pkg-ver,...], "testing":[cat/pkg-ver,...]}
	"""
	if cpvs is None:
		cpvs = VARDB.cpv_all()
	keyword_users = {}
	for cpv in cpvs:
		if cpv.startswith("virtual"):
			continue
		if use_portage:
			keyword = analyser.get_inst_keyword_cpv(cpv)
		else:
			pkg = Package(cpv)
			keyword = analyser.get_inst_keyword_pkg(pkg)
		#print "returned keyword =", cpv, keyword, keyword[0]
		key = keyword[0]
		if key in ["~", "-"]:
			_kwd = keyword[1:]
			if _kwd in keyword_users:
				if key in ["~"]:
					keyword_users[_kwd]["testing"].append(cpv)
				elif key in ["-"]:
					#print "adding cpv to missing:", cpv
					keyword_users[_kwd]["missing"].append(cpv)
			else:
				if key in ["~"]:
					keyword_users[_kwd] = {"stable": [],
						"testing": [cpv], "missing": []}
				elif key in ["-"]:
					keyword_users[_kwd] = {"stable": [],
						"testing": [], "missing": [cpv]}
				else:
					keyword_users[_kwd] = {"stable": [cpv],
						"testing": [], "missing": []}
		elif keyword in keyword_users:
				keyword_users[keyword]["stable"].append(cpv)
		else:
				keyword_users[keyword] = {"stable": [cpv], "testing": [], "missing": []}
	return keyword_users


class Analyse(ModuleBase):
	"""Installed db analysis tool to query the installed databse
	and produce/output stats for USE flags or keywords/mask.
	The 'rebuild' action output is in the form suitable for file type output
	to create a new package.use, package.keywords, package.unmask
	type files in the event of needing to rebuild the
	/etc/portage/* user configs
	"""
	def __init__(self):
		ModuleBase.__init__(self)
		self.module_name = "analyse"
		self.options = {
			"flags": False,
			"keywords": False,
			"unset": False,
			"verbose": False,
			"quiet": False,
			'prefix': False,
			'portage': True
		}
		self.module_opts = {
			"-f": ("flags", "boolean", True),
			"--flags": ("flags", "boolean", True),
			"-k": ("keywords", "boolean", True),
			"--keywords": ("keywords", "boolean", True),
			"-u": ("unset", "boolean", True),
			"--unset": ("unset", "boolean", True),
			"-v": ("verbose", "boolean", True),
			"--verbose": ("verbose", "boolean", True),
			"-p": ("prefix", "boolean", True),
			"--prefix": ("prefix", "boolean", True),
			"-G": ("portage", "boolean", False),
			"--portage": ("portage", "boolean", False),
		}
		self.formatted_options = [
			("    -h, --help",  "Outputs this useage message"),
			("    -a, --analyse",
			"Action, sets the module to gather data and output the"),
			("", "formatted stats/information to the screen"),
			("    -u, --unset",
			"Additionally include any unset USE flags and the packages"),
			("", "that could use them"),
			("    -v, --verbose",
			"Used in the analyse action to output more detailed information"),
			("    -p, --prefix",
			"Used for testing purposes only, runs report using " +
			"a prefix keyword and 'prefix' USE flag"),
			#("    -G, --portage",
			#"Use portage directly instead of gentoolkit's Package " +
			#"object for some operations. Usually a little faster."),
		]
		self.formatted_args = [
			("    use",
			"causes the action to analyse the installed packages USE flags"),
			("    pkguse",
			"causes the action to analyse the installed packages PKGUSE flags"),
			("    ",
			"These are flags that have been set in /etc/portage/package.use"),
			("    keywords",
			"causes the action to analyse the installed packages keywords"),
		]
		self.short_opts = "huvpG"
		self.long_opts = ("help", "unset", "verbose", "prefix") #, "portage")
		self.need_queries = True
		self.arg_spec = "Target"
		self.arg_options = ['use', 'pkguse','keywords']
		self.arg_option = False
		self.warning = (
			"     CAUTION",
			"This is beta software and some features/options are incomplete,",
			"some features may change in future releases includig its name.",
			"Feedback will be appreciated, http://bugs.gentoo.org")


	def run(self, input_args, quiet=False):
		"""runs the module

		@param input_args: input arguments to be parsed
		"""
		query = self.main_setup(input_args)
		query = self.validate_query(query)
		if query in ["use", "pkguse"]:
			self.analyse_flags(query)
		elif query in ["keywords"]:
			self.analyse_keywords()

	def analyse_flags(self, target):
		"""This will scan the installed packages db and analyse the
		USE flags used for installation and produce a report on how
		they were used.

		@type target: string
		@param target: the target to be analysed, one of ["use", "pkguse"]
		"""
		system_use = portage.settings["USE"].split()
		self.printer = AnalysisPrinter("use", self.options["verbose"], system_use)
		if self.options["verbose"]:
			cpvs = VARDB.cpv_all()
			#cpvs = get_installed_cpvs()
			#print "Total number of installed ebuilds =", len(cpvs)
			flag_users = gather_flags_info(cpvs, system_use,
				self.options["unset"], target=target.upper(),
				use_portage=self.options['portage'])
		else:
			cpvs = get_installed_cpvs()
			flag_users = gather_flags_info(cpvs, system_flags=system_use,
				include_unset=self.options["unset"], target=target.upper(),
				use_portage=self.options['portage'])
		#print flag_users
		flag_keys = sorted(flag_users)
		if self.options["verbose"]:
			print("    Flag                              System  #pkgs   cat/pkg-ver")
			blankline = nl
		elif not self.options['quiet']:
			print("    Flag                              System  #pkgs")
			blankline = lambda: None
		for flag in flag_keys:
			flag_pos = flag_users[flag]["+"]
			if len(flag_pos):
				self.printer(flag, "+", flag_pos)
				#blankline()
			flag_neg = flag_users[flag]["-"]
			if len(flag_neg):
				self.printer(flag, "-", flag_neg)
				#blankline()
			if "unset" in flag_users[flag] and flag_users[flag]["unset"]:
				flag_unset = flag_users[flag]["unset"]
				self.printer(flag, "unset", flag_unset)
			#blankline()
		if not self.options['quiet']:
			print("===================================================")
			print("Total number of flags in report =", pp.output.red(str(len(flag_keys))))
			if self.options["verbose"]:
				print("Total number of installed ebuilds =", pp.output.red(str(len([x for x in cpvs]))))
			print()


	def analyse_keywords(self, keywords=None):
		"""This will scan the installed packages db and analyse the
		keywords used for installation and produce a report on them.
		"""
		print()
		system_keywords = portage.settings["ACCEPT_KEYWORDS"]
		arch = portage.settings["ARCH"]
		if self.options["prefix"]:
			# build a new keyword for testing
			system_keywords = "~" + arch + "-linux"
		if self.options["verbose"] or self.options["prefix"]:
			print("Current system ARCH =", arch)
			print("Current system ACCEPT_KEYWORDS =", system_keywords)
		system_keywords = system_keywords.split()
		self.printer = AnalysisPrinter("keywords", self.options["verbose"], system_keywords)
		self.analyser = KeywordAnalyser( arch, system_keywords, VARDB)
		#self.analyser.set_order(portage.settings["USE"].split())
		# only for testing
		test_use = portage.settings["USE"].split()
		if self.options['prefix'] and 'prefix' not in test_use:
			print("ANALYSE_KEYWORDS() 'prefix' flag not found in system USE flags!!!  appending for testing")
			print()
			test_use.append('prefix')
		self.analyser.set_order(test_use)
		# /end testing

		if self.options["verbose"]:
			cpvs = VARDB.cpv_all()
			#print "Total number of installed ebuilds =", len(cpvs)
			keyword_users = gather_keywords_info(
				cpvs=cpvs,
				system_keywords=system_keywords,
				use_portage=self.options['portage'],
				keywords=keywords, analyser = self.analyser
				)
			blankline = nl
		else:
			keyword_users = gather_keywords_info(
				system_keywords=system_keywords,
				use_portage=self.options['portage'],
				keywords=keywords,
				analyser = self.analyser
				)
			blankline = lambda: None
		#print keyword_users
		keyword_keys = sorted(keyword_users)
		if self.options["verbose"]:
			print(" Keyword               System  #pkgs   cat/pkg-ver")
		elif not self.options['quiet']:
			print(" Keyword               System  #pkgs")
		for keyword in keyword_keys:
			kwd_stable = keyword_users[keyword]["stable"]
			if len(kwd_stable):
				self.printer(keyword, " ", kwd_stable)
				blankline()
			kwd_testing = keyword_users[keyword]["testing"]
			if len(kwd_testing):
				self.printer(keyword, "~", kwd_testing)
				blankline()
			kwd_missing = keyword_users[keyword]["missing"]
			if len(kwd_missing):
				self.printer(keyword, "-", kwd_missing)
				blankline
		if not self.options['quiet']:
			if self.analyser.mismatched:
				print("_________________________________________________")
				print(("The following packages were found to have a \n" +
					"different recorded ARCH than the current system ARCH"))
				for cpv in self.analyser.mismatched:
					print("\t", pp.cpv(cpv))
			print("===================================================")
			print("Total number of keywords in report =", pp.output.red(str(len(keyword_keys))))
			if self.options["verbose"]:
				print("Total number of installed ebuilds =", pp.output.red(str(len(cpvs))))
			print()


def main(input_args):
	"""Common starting method by the analyse master
	unless all modules are converted to this class method.

	@param input_args: input args as supplied by equery master module.
	"""
	query_module = Analyse()
	query_module.run(input_args, gentoolkit.CONFIG['quiet'])

# vim: set ts=4 sw=4 tw=79:
